### PR TITLE
Crawl

### PR DESCRIPTION
**Pull request by casperdcl**
Initial proof-of-concept that OpenWorm repo metadata can be scraped.

Intermediate output [CircleCI](https://circleci.com/gh/openworm/openworm_docs/tree/crawl): https://641-12814245-gh.circle-artifacts.com/0/home/circleci/repo/site/gsod19/repos/index.html

- part of #55
- complements #61
- depends on https://github.com/PyGithub/PyGithub/pull/1270
- depends on https://github.com/igraph/python-igraph/pull/255

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+def repo2meta(repo):
+    content = repo2content(repo)
+    if not content:
+        return {}
+    meta = {}
+    for file, data in content.items():
+        if file.endswith('.yml') or file.endswith('.yaml'):
+            meta.update(yaml.safe_load(data))
+        elif file.endswith('.rst'):
```
Comment (mwatts15) [inline build.py]: I have been wondering about this, so I'll ask here: what's the need to support multiple file types? Adding options adds complexity, but I don't see the utility: There's no prior work to integrate, so you can just pick the easiest format to work with (in your assessment).

**Comment by casperdcl** [comment]
I was thinking that we have to make life as easy as possible for subrepo
maintainers, which probably means supporting any format they are used to.

>

**Comment by mwatts15** [comment]
My understanding is that the intention here is to support a couple of paragraphs and an image or two. Markdown is fine for that, whatever your preferred format. For anything more complex, we would still have each sub-project's own documentation site to link to.

**Comment by casperdcl** [comment]
Yes perhaps I'm thinking too long term and big picture.

I was thinking about supporting the current `shortDescription` as well as other lengths (`longDescription`) etc. For projects which are python based and thus have `README.rst` it would be easiest to copy paste bits from.

Happy to go with just markdown for now though - other formats can be a future enhancement if needed. I think the current `NotImplementedError` can be left alone

**Comment by casperdcl** [comment]
btw @mwatts15 any idea why CI isn't reporting status on this PR even though the build is being triggered and is passing? (https://circleci.com/gh/openworm/openworm_docs/tree/crawl)

The last PR (#57) had status reported in the PR.

File: build.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        - current version: {latest_release_date} {latest_release}
+        - contact: <{coordinator}>
+
+        {shortdescription}
+
+        <small>Last generated {latest_generated_date:%Y-%m-%d}</small>
+        """
+            )
+            .format(**fmt)
+            .replace(" | [docs]()", "")
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply build.py]: ah yes ofc readthedocs would support jinja...

[the remark this answers is not in the document]
Comment (casperdcl) [reply build.py]: or would it be better to install jinja locally as a python module and pre-process in-place?

[the remark this answers is not in the document]
Comment (mwatts15) [reply build.py]: Ah. I was thinking of using Jinja in place of the .format() call. Normally I would applaud avoiding a dependency, but I figure Jinja is a bit less brittle.

**Comment by mwatts15** [comment]
> btw @mwatts15 any idea why CI isn't reporting status on this PR even though the build is being triggered and is passing? (https://circleci.com/gh/openworm/openworm_docs/tree/crawl)
> 
> The last PR (#57) had status reported in the PR.

This is OK:
image

**Comment by casperdcl** [comment]
regarding spec (i.e. supported keys) in `.openworm.yml`:

> @Casper.dcl Apologies for the late response. I do not believe we have a formal spec for how `.openworm.yml` is setup. I recall it was a little adhoc, and had a mechanism where unrecognized terms were ignored.
>
> Perhaps for this project we should use something else, say `.documentation.yml`? We should probably discuss this with @balicea later.

-- @cheelee

**Comment by casperdcl** [comment]
So for now I just crawled through all the current `.openworm.yml` keys and found:
https://github.com/openworm/openworm_docs/blob/8648d0ff8a1603ef517e899b3007da180ea18854/build.py#L83-L106

This would need documentation I'm sure. Also there are a number of issues with it.

- what was `shortDescription` intended for? As of now, if it's blank, I use the repository's description from GitHub - i.e. in the image below, the description is "Documentation of OpenWorm," while the homepage is "http://docs.openworm.org"
image
I guess it's not a big deal to have it duplicated but we need to be explicit about which overrides which if both are defined
- in the image above you can also click "manage topics" which could similarly clash with `keywords`


clashing keys:

`.openworm.yml` | GitHub
--:|:--
members|contributors
parent|parent (?)
contributor_guide|(.github/)CONTRIBUTING(.md/rst/txt)
coordinator|max(collaborators, key=commit.count)
documentation|readme
keywords|topics
languages|languages
latest_release, latest_release_date|latest_release, latest_tag, latest_tag prefixed with `v/V`
repo|name, full_name, url, git_url, ssh_url
shortDescription|description
latest_generated_date (new in this PR)|openworm.yml.last_modified

There's actually a HUGE list of metadata that I can get for repos from the GitHub API. A small relevant subset would be:

- archives
- clones
- collaborators
- contributors
- created_at
- default_branch
- description
- downloads
- forks
- full_name
- url (html, git-html, git-ssh, homepage, as well as downloads, forks, issues, stargazers, etc)
- issues
- languages
- last_modified
- latest_release
- license
- name
- parent
- projects
- pulls
- pushed_at
- raw_data
- readme
- releases
- size
- source
- stargazers
- stats_code_frequency
- stats_commit_activity
- stats_participation
- stats_punch_card
- subscribers
- tags
- teams
- topics
- updated_at
- views_traffic
- watchers
- wiki

**Comment by casperdcl** [comment]
@mwatts15 re: CI reporting on GitHub, I was referring to:

image

Where there's a tick next to the commit. Maybe it'll be easiest if you can grant me admin permissions to this repo so I can check the webhooks etc. The description at the top of this PR includes a badge which I had to manually add. Ideally instead circle-ci build stats should show up automatically at the bottom next to the merge button.

File: build.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (casperdcl) [reply build.py]: I agree it's less brittle but also given the sheer number of templates a little less elegant too. idk.
https://github.com/openworm/openworm_docs/blob/8648d0ff8a1603ef517e899b3007da180ea18854/docs/gsod19/repos.md.template is pretty elegant without jinja syntax. Also the python `{}` would clash with jinja `{}` so we'd have to use even quadruple `{{{{}}}}` in some places. Using `%` in python would also potentially clash with jinja.

**Comment by casperdcl** [comment]
@slarson @MichaelCurrie @cheelee @balicea on the subject of #32: I was playing with https://igraph.org/python/ and the metadata from this PR and have this:

repos

**Comment by casperdcl** [comment]
Finally (slightly) better interactive graph builds... https://641-12814245-gh.circle-artifacts.com/0/home/circleci/repo/site/gsod19/repos/index.html

File: build.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwatts15) [reply build.py]: I don't understand your argument: I was suggesting using Jinja in place of `str.format` or `%` format, so the clashes you mention would be irrelevant. Jinja also lets you put the case that you don't have a datum for the project directly in the template so you don't need these calls to `replace` after substitution into the template.

**Comment by mwatts15** [comment]
I see the merge conflict -- otherwise, what's the status of this PR? The repositories page you linked above looks OK except for YAROM looks like it's missing some newlines.

File: build.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (casperdcl) [reply build.py]: sure, yes, jinja in-place. Maybe for a different PR? This one's getting a bit complex already.

**Comment by casperdcl** [comment]
@mwatts15 https://github.com/openworm/YAROM/blob/dev/.openworm.yml doesn't exactly have anything to go by

**Comment by casperdcl** [comment]
Ah I understand. yes

**Pull request merged**